source/common/http/conn_manager_impl.cc

```diff
@@ -258,10 +258,22 @@ RequestDecoder& ConnectionManagerImpl::newStream(ResponseEncoder& response_encod
   new_stream->response_encoder_ = &response_encoder;
   new_stream->response_encoder_->getStream().addCallbacks(*new_stream);
   new_stream->response_encoder_->getStream().setFlushTimeout(new_stream->idle_timeout_ms_);
-  // If the network connection is backed up, the stream should be made aware of it on creation.
-  // Both HTTP/1.x and HTTP/2 codecs handle this in StreamCallbackHelper::addCallbacksHelper.
-  ASSERT(read_callbacks_->connection().aboveHighWatermark() == false ||
-         new_stream->filter_manager_.aboveHighWatermark());
+
+  // If the network connection is backed up, the HTTP/1.x stream should be made aware of it on
+  // creation. In the case of HTTP/2 the stream should be allowed to read up to the configured
+  // stream limit even when the network connection is backed up, so the readDisable status is not
+  // propagated from the network connection if the
```

Comment:
I may get unconfused by the time I get to the bottom of this PR but

What this used to do was inform new streams to bound their reads if the upstream connection was backed up.
Read disabling for HTTP/2 didn't stop it from reading a window's worth of data, it immediately stopped it from sending window updates to ensure that only one windows's worth of data would be read.

I'm not sure how delaying serialization should affect this - I'd think we'd still want to inform the downstream codec to stop sending window updates if upstream is at capacity.

Reply:
Yes, there's a difference in buffering behavior.  I think flow control ends up operating based on the contents of the stream buffer and depend on readDisable when that buffer is above watermark to suppress window updates.  I think that we could end up buffering more bytes total per connection, but we'ld end up with stream buffers which are 2x the high configured watermark instead of connection buffers with >100x high watermark bytes in them.

There's some issues that I haven't managed to work out yet, like the configured buffer sizes for the downstream stream buffer affecting how much we download from the upstream over H1 or H2.  The issue of H2 stream buffers now playing a role in cases where the other end of the HTTP handler is using H1 could cause some problems related to default values in the config: A downstream H2 connection without explicitly configured defaults or buffer limits would buffer 1MB in the connection's output buffer when proxying data from an H1 upstream.  After this change, the buffering would be done based on the H2 stream window, so we'ld buffer 1MB in the connection's output buffer and another 256MB in the stream's buffer since that' the default.  This behavior is illustrated by Http2BufferWatermarksTest.DataFlowControlHttp1Upstream

We may want to consider keeping the runtime features associated with this change disabled by default and only consider enabling them after doing several followup fixes.

Reply:
one thing that might help this PR is updating flow_control.md to explain what's going on?  Maybe that'd help make sure I grokked what you were aiming for and can match it up with what the PR does?

Reply:
Sorry for not replying until now.  It's been hard for me to find time to continue doing the doc changes required for this.

One thing that we may want to answer is wherever or not the runtime feature knob added in this PR should default to true or false.  There are some undesirable behaviors related to buffering when one side of the pipeline is H1 and the other is H2, but the extra buffering does reduce changes for starvation.  I need to write a proposal for changes to how H2 pushes the pipeline which would reduce buffering in several interesting cases and address several other issues.  If we decide to hold back the runtime feature, should we also hold back documentation?  I think we do want new documentation to be ready when we merge functionality with the flag off, commit to getting remaining issues sorted out in the short term and switch over to delayed serialization soon.